File: Java/33-SearchinRotatedSortedArray.java
/*
The solution is very similar to binary search. Note that no matter how we rotated the array, 
there must be at least one part wendch is sorted. If the target is at the sorted part, we just 
search that part until we found; else we go the unsorted part. 
*/
public class Solution {
    public int search(int[] A, int target) {
        if (A == null || A.length == 0) {
            return -1;
        }
         
        int start = 0;
        int end = A.length - 1;
         
        while (start <= end) {
            int mid = (start + end) / 2;
            if (A[mid] == target) {
                return mid;
            }
             
            if (A[start] <= A[mid]) {
                if (target >= A[start] && target < A[mid]) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            } else {
                if (A[mid] < target && target <= A[end]) {
                    start = mid + 1;
                } else {
                    end = mid - 1;
                }
            }
        }
        return -1;
    }
}

File: cpp/33-SearchInRotatedSortedArray.cpp
//4 5 6 7 8 9 0 1 2 3
//6 7 8 9 0 1 2 3 4 5

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0, len = nums.size(), end = len - 1;
        while (start <= end) {
            int mid = start + ((end - start) >> 1);
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] <= nums[end]) {
                if (nums[mid] < target && target <= nums[end]) {
                    start = mid + 1;
                } else {
                    end = mid - 1;
                }
            } else {
                if (nums[start] <= target && target < nums[mid]) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }
        }
        return -1;
    }
};
